.ci/wercker.sh

```diff
@@ -234,6 +234,19 @@ no-error-spring-integration)
   removeFolderWithProtectedFiles spring-integration
   ;;
 
+no-error-spring-cloud-gcp)
+  set -e
+  CS_POM_VERSION=$(mvn -e -q -Dexec.executable='echo' -Dexec.args='${project.version}' \
+                     --non-recursive org.codehaus.mojo:exec-maven-plugin:1.3.1:exec)
+  echo CS_version: ${CS_POM_VERSION}
+  checkout_from https://github.com/spring-cloud/spring-cloud-gcp
+  cd .ci-temp/spring-cloud-gcp
+  mvn -e compile checkstyle:check -Dproject.version=${CS_POM_VERSION} \
+     -Dcheckstyle.configLocation=../../config/checkstyle_checks.xml
```

Comment:
we should use config from spring-cloud-gcp, `checkstyle_checks.xml` is only applicable to our repo ONLY, it is too strict to be used in other projects.

Reply:
@elefeint  , please share a CI command on how to run only checkstyle validation on `spring-cloud-gcp`. We should use exactly the same command as in your CI.

Reply:
Alright, i’ll change this. Also should I use mvn or mvnw? spring-cloud-gcp has provided mvnw command in the issue

Reply:
Either `mvn` or `mvnw` should work. I have maven 3.6.2 locally, which works.

Try omitting the `-Dcheckstyle.configLocation` override from your invocation. The project's pom file should find the right configuration by default.

Reply:
Alright, I’ll try this. Thanks 👍

Reply:
from wercker log:
```
[INFO] --- maven-checkstyle-plugin:3.0.0:check (default-cli) @ spring-cloud-gcp-core ---
[INFO] There are 135 errors reported by Checkstyle 8.18 with sun_checks.xml ruleset.
```
attention to `sun_checks.xml` it is clear problem with execution command.
@Gaurav-Punjabi , from where you get this command ? it looks like copy-paste from somewhere, very unrelated.

> -Dproject.version=${CS_POM_VERSION}

we need to have property in spring-cloud-gcp-core project to let us specify SNAPSHOT version

Reply:
@elefeint , looks like checkstyle version defined somewhere in parents, is it possible to have for us variable in pom.xml that we can specify during build execution to specify our SNAPSHOT version of checkstyle.

Reply:
@romani I got the command from equals verifier config. Also I ran `./mvnw checkstyle:check@checkstyle-validation ` without specifying the location to config file. It generated 25 violations.

Reply:
I looked into it more; the problem is in the difference between maven invoking a phase (as we do in our build) and invoking a specific goal (as your CI is doing). The phase invocation `mvn validate` would correctly configure the checkstyle execution named `checkstyle-validation` (we can't rename that because it needs to override the global Spring Cloud configuration). But the goal invocation `mvn checkstyle:check` creates an execution with fresh defaults, which are to run `sun_checks.xml`, which our project does not abide by.

So there are two ways to invoke checkstyle that will work for Spring Cloud GCP:
_Option1 (invoke goal and provide the execution configuration name):_
`mvn checkstyle:check@checkstyle-validation -P full-checkstyle`

_Option2 (invoke phase and let maven configure the correct goal):_
`mvn validate -P full-checkstyle`

(`-P full-checkstyle` enables checkstyle in our sample applications; it's nice to have for Option 2 and necessary for Option1 -- otherwise the default checkstyle config with `sun_checks.xml` will be in effect)

Reply:
@elefeint It worked! 👍 Thanks. One last thing, as specified by roman in this comment https://github.com/checkstyle/checkstyle/pull/8106#discussion_r413029362. Is there a way we can specify checkstyle snapshot version during build execution?

Reply:
Do you really need `-Dproject.version=${CS_POM_VERSION}` if you are just running from root? The project version is embedded in the checked out pom file; it does not need to be overridden.

Reply:
oh, I understand what you mean; to test the latest version of checkstyle plugin, you need to override the one we have. 

Use `maven-checkstyle-plugin.version`. It will work on everything except samples project, which overrides it for some reason. I'll remove the override on our side.

Reply:
So, `-Dmaven-checkstyle-plugin.version=${CS_POM_VERSION}`

Reply:
Actually, that's wrong; that one is for the maven plugin. Use `puppycrawl-tools-checkstyle.version` for the tool itself.

Reply:
@romani I tried this command provided by @elefeint 
```
 mvn validate -P full-checkstyle -Dpuppycrawl-tools-checkstyle.version=8.32-SNAPSHO
```

But it is throwing this error
```
[ERROR] Failed to execute goal org.apache.maven.plugins:maven-checkstyle-plugin:3.0.0:check (checkstyle-validation) on project spring-cloud-gcp: Execution checkstyle-validation of goal org.apache.maven.plugins:maven-checkstyle-plugin:3.0.0:check failed: An API incompatibility was encountered while executing org.apache.maven.plugins:maven-checkstyle-plugin:3.0.0:check: java.lang.NoSuchMethodError: com.puppycrawl.tools.checkstyle.DefaultLogger.<init>(Ljava/io/OutputStream;Z)V
```

It would be great if you could just help me this. I'm stuck here.

Reply:
Maven checkstyle plugin is [still using](https://github.com/apache/maven-checkstyle-plugin/blob/36bde8d75aba5e109637bb0762965f76fbb027b2/src/main/java/org/apache/maven/plugins/checkstyle/CheckstyleViolationCheckMojo.java#L751) the `DefaultLogger` constructors that were [removed](https://github.com/checkstyle/checkstyle/commit/06ee21624f35546412b260e5d8c55d906dda92e4#diff-ecc827b26f47a5f0023c5ae4a0871d63) in checkstyle.

You'll have to work with the [maven-checkstyle-plugin](https://github.com/apache/maven-checkstyle-plugin) project to get them to stop using nonexistent constructors.

Reply:
> maven-checkstyle-plugin:3.0.0:check

@elefeint , is it possible to upgrade maven plugin to 3.1.1 in spring-cloud-gcp ?
http://maven.apache.org/plugins/maven-checkstyle-plugin/
it is well known problem that we already resolved in 3.1.1 version.
-----

@Gaurav-Punjabi , please use `mvn checkstyle:check@checkstyle-validation -P full-checkstyle` we need to minimize execution time as much as possible.

Reply:
> @Gaurav-Punjabi , please use `mvn checkstyle:check@checkstyle-validation -P full-checkstyle` we need to minimize execution time as much as possible.

Yeah, I was going to use this command. I was just testing it with `mvn validate 👍

Reply:
Upgraded to the latest maven plugin and checkstyle in spring-cloud/spring-cloud-gcp#2337. Try it now?

Reply:
@elefeint I tried running the command again, the dependency is updated but it is throwing this error.
```
[ERROR] Failed to execute goal org.apache.maven.plugins:maven-checkstyle-plugin:3.1.1:check (checkstyle-validation) on project spring-cloud-gcp: Execution checkstyle-validation of goal org.apache.maven.plugins:maven-checkstyle-plugin:3.1.1:check failed: Plugin org.apache.maven.plugins:maven-checkstyle-plugin:3.1.1 or one of its dependencies could not be resolved: Failure to find com.puppycrawl.tools:checkstyle:jar:8.32 in https://repo.spring.io/libs-milestone-local was cached in the local repository, resolution will not be reattempted until the update interval of spring-milestones has elapsed or updates are forced
```

I tried re-installing all the dependencies, still encountering this issue.

Reply:
@elefeint `DefaultLogger` issue is resolved. Just one last thing, the checkstyle config for `spring-cloud-gcp-samples` has a property `allowMissingJavadoc` in `JavadocMethodCheck`. This property doesn't exist in the check as you can see [here](https://checkstyle.sourceforge.io/config_javadoc.html#JavadocMethod_Properties). Can you please remove that property from the config ?

You can use this command to check this issue
```
 mvn validate -P full-checkstyle -Dpuppycrawl-tools-checkstyle.version=8.32-SNAPSHOT -Dmaven-checkstyle-plugin.version=3.1.1
``` 

Here's the error log for your reference :)
```
[ERROR] Failed to execute goal org.apache.maven.plugins:maven-checkstyle-plugin:3.1.1:check (checkstyle-validation) on project spring-cloud-gcp-samples: Failed during checkstyle configuration: cannot initialize module com.puppycrawl.tools.checkstyle.TreeWalker - cannot initialize module com.puppycrawl.tools.checkstyle.checks.javadoc.JavadocMethodCheck - Property 'allowMissingJavadoc' does not exist, please check the documentation -> [Help 1]
```

Reply:
Just merged a fix. Try it again?

Reply:
It works and we're good to go. I'll make the changes and will soon get it reviewed by the members. Thanks for your corporation 👍🏻

Reply:
Please use `mvn checkstyle:check@checkstyle-validation -P full-checkstyle` to execute what we need only

Reply:
Oh yeah my bad, Fixed it 👍